sklearn/datasets/value_dropper.py

```diff
@@ -0,0 +1,394 @@
+           [ nan,  nan,   3.]])
+    """
+
+    def __init__(self, missing_values="NaN",
+                 missing_distribution=None, copy=True, random_state=None):
+        self.missing_values = missing_values
+        self.missing_distribution = missing_distribution
+        self.copy = copy
+        self.random_state = random_state
+
+    def transform(self, X, y=None):
+        """Drop values from ``X`` according to the given distribution.
+
+        Parameters
+        ----------
+
+        X : ndarray like of shape (n_features, n_samples)
+            Data, in which the values must be dropped and set to
+            ``missing_values``.
+
+        y : array-like, shape = (n_samples,), optional for MCAR
+            Target relative to X for classification or regression;
+            When missing_distribution is not a dict (for MCAR missingness),
+            ``y`` need not be passed.
+        """
+        # Validate missing_values and generate missing_mask
+        if ((isinstance(self.missing_values, str) and
+                (self.missing_values.lower() == "nan")) or
+                np.isnan(self.missing_values)):
+            missing_values = np.nan
+        else:
+            missing_values = self.missing_values
+
+        # Don't allow pre-exising missing values in X, to simplify API
+        X = check_array(X, dtype=('numeric'
+                                  if isinstance(missing_values,
+                                                (numbers.Integral, np.integer))
+                                  else np.float),
+                        copy=self.copy)
+
+        n_samples, n_features = X.shape
+        n_values = n_samples * n_features
+
+        rng = check_random_state(self.random_state)
+
+        # Validate y, and find type of missingness
+        if isinstance(self.missing_distribution, dict):
+            missing_type = 'nmar'
+            # For NMAR
+            # Validate and convert the missing_distribution dict into a
+            # 2D probability distribution along the features and labels
+
+            if y is None:
+                raise ValueError("The missing_distribution is a dict "
+                                 "but y is None. If missingness is to be "
+                                 "related to the class labels, target class "
+                                 "labels (y) must be passed.")
+
+            target_type = type_of_target(y)
+            if 'continuous' in target_type or 'multioutput' in target_type:
+                raise ValueError("Value dropping based on the given "
+                                 "distribution can be done only for single "
+                                 "target which is discrete (classification "
+                                 "tasks). The given target (y) is of type %s"
+                                 % target_type)
+            y = check_array(y, ensure_2d=False, dtype='numeric')
+
+            le = LabelEncoder().fit(y)
+            classes = le.classes_
```

Comment:
You mean combining both lines into one? (I didn't because I need the `le` later on (for encoding the keys of the `missing_distribution`)